src/fetch_blocks.rs

```diff
@@ -0,0 +1,395 @@
+                .await?;
+            match msg.payload {
+                NetworkMessage::Block(b) => {
+                    let returned_hash = b.block_hash();
+                    let merkle_check = b.check_merkle_root();
+                    let witness_check = b.check_witness_commitment();
+                    return match (returned_hash == hash, merkle_check, witness_check) {
+                        (true, true, true) => {
+                            peer.mark_clean();
+                            Ok(b)
+                        }
+                        (true, true, false) => {
+                            Err(anyhow::anyhow!("Witness check failed for {:?}", hash))
+                        }
+                        (true, false, _) => {
+                            Err(anyhow::anyhow!("Merkle check failed for {:?}", hash))
+                        }
+                        (false, _, _) => Err(anyhow::anyhow!(
+                            "Expected block hash {:?}, got {:?}",
+                            hash,
+                            returned_hash
+                        )),
+                    };
+                }
+                NetworkMessage::Ping(p) => {
+                    peer.with_connection_blocking(env.clone(), move |conn| {
+                        RawNetworkMessage {
+                            magic: Bitcoin.magic(),
+                            payload: NetworkMessage::Pong(p),
+                        }
+                        .consensus_encode(conn)
+                        .map_err(From::from)
+                    })
+                    .await?;
+                }
+                m => warn!(env.logger, "Invalid Message Received: {:?}", m),
+            }
+        }
+    })
+    .await?
+}
+
+async fn fetch_block_from_peer_set(
+    env: Arc<Env>,
+    peers: Vec<PeerHandle>,
+    hash: BlockHash,
+) -> Option<Block> {
+    use futures::stream::StreamExt;
+
+    let (send, mut recv) = futures::channel::mpsc::channel(1);
+    let fut_unordered: futures::stream::FuturesUnordered<_> =
+        peers.into_iter().map(futures::future::ready).collect();
+    let env_local = env.clone();
+    let runner = fut_unordered
+        .then(move |p| fetch_block_from_peer(env_local.clone(), hash.clone(), p))
+        .for_each_concurrent(None, |block_res| {
+            match block_res {
+                Ok(block) => send.clone().try_send(block).unwrap_or_default(),
+                Err(e) => warn!(env.logger, "Error fetching block from peer: {}", e),
+            }
+            futures::future::ready(())
+        });
+    let b = futures::select! {
+        b = recv.next().fuse() => b,
+        _ = runner.boxed().fuse() => None,
+    };
+    b
+}
+
+pub async fn fetch_block(
+    env: Arc<Env>,
+    peers: Vec<PeerHandle>,
+    hash: BlockHash,
+) -> Result<Option<Block>, RpcError> {
+    Ok(match fetch_block_from_self(&*env, hash).await? {
+        Some(block) => Some(block),
+        None => {
+            info!(
```

Comment:
I suggest decreasing to `debug`. Otherwise if you run `electrs` against this, you'll get a full log of useless messages.